From 78817d34a1158843676b91d13f962b270497c813 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Thu, 18 Sep 2014 14:14:19 -0400
Subject: [scsi] megaraid_sas: Use writeq for 64bit pci write to avoid spinlock overhead

Message-id: <1411049668-28841-20-git-send-email-thenzl@redhat.com>
Patchwork-id: 96157
O-Subject: [RHEL7.1 e-stor PATCH 19/28] megaraid_sas : Use writeq for 64bit pci write to avoid spinlock overhead
Bugzilla: 1088523
RH-Acked-by: Ewan Milne <emilne@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

Resending the patch. Addressed the review comments from Tomas Henzl.
Reduce the assingment for u64 req_data variable.

Use writeq() for 64bit PCI write instead of writel() to avoid additional lock overhead.

scsi-queue/drivers-for-3.18 commit 0756040952582c4e7c2f23ff2af882c0f0c7a516

Signed-off-by: Sumit Saxena <sumit.saxena@avagotech.com>
Signed-off-by: Kashyap Desai <kashyap.desai@avagotech.com>
Reviewed-by: Tomas Henzl <thenzl@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/megaraid/megaraid_sas_fusion.c b/drivers/scsi/megaraid/megaraid_sas_fusion.c
index 0858851..dd983d7 100644
--- a/drivers/scsi/megaraid/megaraid_sas_fusion.c
+++ b/drivers/scsi/megaraid/megaraid_sas_fusion.c
@@ -1065,6 +1065,11 @@ megasas_fire_cmd_fusion(struct megasas_instance *instance,
    u32 req_desc_hi,
    struct megasas_register_set __iomem *regs)
 {
+#if defined(writeq) && defined(CONFIG_64BIT)
+ u64 req_data = (((u64)req_desc_hi << 32) | (u32)req_desc_lo);
+
+ writeq(le64_to_cpu(req_data), &(regs)->inbound_low_queue_port);
+#else
  unsigned long flags;
 
  spin_lock_irqsave(&instance->hba_lock, flags);
@@ -1072,6 +1077,7 @@ megasas_fire_cmd_fusion(struct megasas_instance *instance,
  writel(le32_to_cpu(req_desc_lo), &(regs)->inbound_low_queue_port);
  writel(le32_to_cpu(req_desc_hi), &(regs)->inbound_high_queue_port);
  spin_unlock_irqrestore(&instance->hba_lock, flags);
+#endif
 }
 
 /**
-- 
1.7.1